IMDB Movie dataset is a movie review dataset, used to analyze the interest level of movies according to a number of criteria 
such as: director, actors, movie name... to provide perspectives to support prediction. in the future.

# 1. Import libraries and load dataset: 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
dataset_path = "./data/IMDB-Movie-Data.csv"

In [3]:
# Read data from .csv file
data = pd.read_csv(dataset_path)

In addition, we can read and specify a column as an index for the data table (by default pandas will create a separate 
index column). Here, we can choose the Title column as the index column as follows (the index column cannot 
contain duplicate values):

In [4]:
# Read data with specified explicit index .
# We will use this later in our analysis
data_indexed = pd.read_csv (dataset_path,index_col ="Title")

In [5]:
data_indexed.head()

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
Split                                138.12       62.0  
Sing                                 270.32       59.0  
Suicide Squad                        325.02       40.0

# 2. View the data: Look through the first 5 rows of the data table using head()

In [6]:
# Preview top 5 rows using head ()
data.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

# 3. Understand some basic information about the data:

In [7]:
# Let 's first understand the basic information about this data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


In [8]:
data . describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)   Metascore  
count          872.000000  936.000000  
mean            82.956376   58.985043  
std            103.253540   17.194757  
min              0.000000   11.000000  
25%             13.270000   47.000000  
50%             47.985000   59.500000  
75%            113.715000   72.000000  
max            936.630000  100.000000

- Min and max values of Year, meaning the dataset contains movies from 2006 to 2016. • 
- The average rating for movies is 6.7, the lowest is 1.9, the highest is 9.0. •
- The highest revenue achieved was 936.6 million dollars

# 4. Data Selection 

- Indexing and Slicing data: From the data table, we can split any column in the data table to 
become a Series or a DataFrame, depending on the splitting method we use. Here, we will separate some 
columns in the data using Indexing technique. To split columns into Series, do:

In [9]:
 # Extract data as series
genre = data["Genre"]
genre.to_frame()

Genre
0     Action,Adventure,Sci-Fi
1    Adventure,Mystery,Sci-Fi
2             Horror,Thriller
3     Animation,Comedy,Family
4    Action,Adventure,Fantasy
..                        ...
995       Crime,Drama,Mystery
996                    Horror
997       Drama,Music,Romance
998          Adventure,Comedy
999     Comedy,Family,Fantasy

[1000 rows x 1 columns]

In [10]:
data.columns

Index(['Rank', 'Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

We can select and split multiple columns at the same time, creating a new DataFrame:

In [11]:
some_col = data[["Title","Genre","Actors","Director", "Rating"]]
some_col.head()

Title                     Genre  \
0  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1               Prometheus  Adventure,Mystery,Sci-Fi   
2                    Split           Horror,Thriller   
3                     Sing   Animation,Comedy,Family   
4            Suicide Squad  Action,Adventure,Fantasy   

                                              Actors              Director  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...            James Gunn   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...          Ridley Scott   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...    M. Night Shyamalan   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  Christophe Lourdelet   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...            David Ayer   

   Rating  
0     8.1  
1     7.0  
2     7.3  
3     7.2  
4     6.2

For row splitting, we can separate a certain number of rows, from index X to index Y in the data table, called 
Slicing. For example, to separate rows 10 through 15, do the following:


In [12]:
data.iloc[10:15][["Title","Genre","Actors","Director", "Rating"]]

Title                       Genre  \
10  Fantastic Beasts and Where to Find Them    Adventure,Family,Fantasy   
11                           Hidden Figures     Biography,Drama,History   
12                                Rogue One     Action,Adventure,Sci-Fi   
13                                    Moana  Animation,Adventure,Comedy   
14                                 Colossal         Action,Comedy,Drama   

                                               Actors         Director  Rating  
10  Eddie Redmayne, Katherine Waterston, Alison Su...      David Yates     7.5  
11  Taraji P. Henson, Octavia Spencer, Janelle Mon...   Theodore Melfi     7.8  
12  Felicity Jones, Diego Luna, Alan Tudyk, Donnie...   Gareth Edwards     7.9  
13  Auli'i Cravalho, Dwayne Johnson, Rachel House,...     Ron Clements     7.7  
14  Anne Hathaway, Jason Sudeikis, Austin Stowell,...  Nacho Vigalondo     6.4

In [13]:
data.columns

Index(['Rank', 'Title', 'Genre', 'Description', 'Director', 'Actors', 'Year',
       'Runtime (Minutes)', 'Rating', 'Votes', 'Revenue (Millions)',
       'Metascore'],
      dtype='object')

Data Selection - Based on Conditional filtering: We can retrieve rows in the data table based on some conditions 
that need to be followed. For example, we want to get movies from 2010 to 2015, with ratings less than 6.0 but 
with revenue in the top 10% of the entire dataset. Accordingly, we can implement the code as follows:

In [14]:
threshold = data["Revenue (Millions)"].quantile(0.9)
filtered_data = data[((data["Year"] > 2010) & (data["Year"] < 2015)) & (data["Rating"] < 5.0) & (data["Revenue (Millions)"] > threshold)]

In [15]:
filtered_data

Rank                                      Title                    Genre  \
925   926  The Twilight Saga: Breaking Dawn - Part 1  Adventure,Drama,Fantasy   

                                           Description     Director  \
925  The Quileutes close in on expecting parents Ed...  Bill Condon   

                                                Actors  Year  \
925  Kristen Stewart, Robert Pattinson, Taylor Laut...  2011   

     Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
925                117     4.9  190244              281.28       45.0

# 6. Groupby Operations

Groupby is a grouping of data based on one or more variables (here, data columns in a 
table). For example, we can find the average rating achieved by directors by grouping the ratings of movies 
by Director.

In [16]:
data.groupby("Director")["Rating"].mean().to_frame().sort_values("Rating", ascending = False)

Rating
Director                 
Nitesh Tiwari        8.80
Christopher Nolan    8.68
Olivier Nakache      8.60
Makoto Shinkai       8.60
Aamir Khan           8.50
...                   ...
Micheal Bafaro       3.50
Jonathan Holbrook    3.20
Shawn Burkett        2.70
James Wong           2.70
Jason Friedberg      1.90

[644 rows x 1 columns]

In [17]:
data.groupby("Director")["Votes"].mean().to_frame().sort_values("Votes", ascending = False)

Votes
Director                    
Christopher Nolan  1311817.0
James Cameron       935408.0
Joss Whedon         781241.5
Quentin Tarantino   639896.5
Tim Miller          627797.0
...                      ...
Mark Williams          164.0
Alexi Pappas           115.0
Gillies MacKinnon      102.0
David Leveaux           96.0
Amber Tamblyn           61.0

[644 rows x 1 columns]

In [18]:
data.groupby("Director")["Revenue (Millions)"].sum().to_frame().sort_values("Revenue (Millions)", ascending = False)

Revenue (Millions)
Director                             
J.J. Abrams                   1683.45
David Yates                   1630.51
Christopher Nolan             1515.09
Michael Bay                   1421.32
Francis Lawrence              1299.81
...                               ...
Jalil Lespert                    0.00
Jamal Hill                       0.00
James Franco                     0.00
James Lapine                     0.00
Greg Kwedar                      0.00

[644 rows x 1 columns]

# 7. Sorting Operations: 

Sorting allows us to sort rows in a data table in ascending/descending order based on the 
value of a certain column in the data table. For example, based on the groupby results of the previous 
section, we can find the top 5 directors with the highest average ratings as follows:

In [19]:
data.groupby("Director")["Rating"].mean().to_frame().sort_values("Rating", ascending = False).head()

Rating
Director                 
Nitesh Tiwari        8.80
Christopher Nolan    8.68
Olivier Nakache      8.60
Makoto Shinkai       8.60
Aamir Khan           8.50

# 8. View missing values:

Data sets will often have missing values in some information fields of some data 
samples. When processing data, we need to overcome this problem. Therefore, the first thing we need to 
check is the location of data loss in the following way:

In [20]:
 # To check null values row - wise
data.isnull().sum().to_frame()

0
Rank                  0
Title                 0
Genre                 0
Description           0
Director              0
Actors                0
Year                  0
Runtime (Minutes)     0
Rating                0
Votes                 0
Revenue (Millions)  128
Metascore            64

Here we see Revenue (Millions) and Metascore are two columns containing null data. To handle the 
problem of data loss, there are two main options: either replace empty areas with some value or remove 
them

# 9. Deal with missing values - Deleting

- Deleting For the removal option, we can remove the entire column containing 
many null values (if possible) or only remove rows containing null values. To delete columns, we do:

In [21]:
data.drop("Metascore", axis = 1).head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  
0     8.1  757074              333.13  
1     7.0  485820              126.46  
2     7.3  157606              138.12  
3     7.2   60545              270.32  
4     6.2  393727              325.02

In [22]:
data.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

The above command still does not drop real data on the server until we add inplace=True.

To delete rows, we use

In [23]:
data.dropna().head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

# 10. Dealing with missing values - Filling:

In [24]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


In [25]:
aver_revenue = data["Revenue (Millions)"].mean()
print(f"The average of revenue is {aver_revenue} Millions")

The average of revenue is 82.95637614678898 Millions


In [26]:
# we can fill missing value with this mean revenue
data["Revenue (Millions)"].fillna(aver_revenue, inplace = True)

In [27]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  1000 non-null   float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


# 11. apply() functions: 

Apply functions are used when we want to execute a function on the rows in the data 
table. After execution, the result returned from the main function is the new value of the corresponding row. 
For example, if we want to classify movies into three levels ['Good', 'Average', 'Bad'] based on Rating, we 
can define a function to do this and apply it to the DataFrame:

In [28]:
# Classify movies based on ratings
def classify_rating(rating):
    if rating >= 7.5:
        return "Good"
    elif rating >= 6.0:
        return "Average"
    else:
        return "Bad"

In [29]:
# Lets apply this function on our movies data
# creating a new variable in the dataset to hold the rating category
data["Rating Category"] = data["Rating"].apply(classify_rating)

In [32]:
data[["Title","Genre","Rating","Rating Category"]].head()

Title                     Genre  Rating Rating Category
0  Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1            Good
1               Prometheus  Adventure,Mystery,Sci-Fi     7.0         Average
2                    Split           Horror,Thriller     7.3         Average
3                     Sing   Animation,Comedy,Family     7.2         Average
4            Suicide Squad  Action,Adventure,Fantasy     6.2         Average

 DataFrame after the rating_group() function is applied. The results returned after executing this 
row will be put into a new column named Rating_category